sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -2004,27 +2004,38 @@ pi_result _pi_queue::insertActiveBarriers(pi_command_list_ptr_t &CmdList,
   if (ActiveBarriers.empty())
     return PI_SUCCESS;
 
-  // Create a wait-list and retain events. This will filter out finished events.
+  // Create a wait-list and retain events.
   _pi_ze_event_list_t ActiveBarriersWaitList;
   if (auto Res = ActiveBarriersWaitList.createAndRetainPiZeEventList(
-          ActiveBarriers.size(), ActiveBarriers.data(), this, UseCopyEngine))
+          ActiveBarriers.vector().size(), ActiveBarriers.vector().data(), this,
+          UseCopyEngine))
     return Res;
 
-  // We can now release all the active barriers and replace them with the ones
-  // in the wait list.
-  for (pi_event &BarrierEvent : ActiveBarriers)
-    PI_CALL(piEventReleaseInternal(BarrierEvent));
+  if (ActiveBarriersWaitList.Length == 0) {
+    ActiveBarriers.clear();
+    return PI_SUCCESS;
+  }
+  
+  // We can now replace active barriers with the ones in the wait list.
   ActiveBarriers.clear();
-  ActiveBarriers.insert(
-      ActiveBarriers.end(), ActiveBarriersWaitList.PiEventList,
-      ActiveBarriersWaitList.PiEventList + ActiveBarriersWaitList.Length);
+  for (pi_uint32 I = 0; I < ActiveBarriersWaitList.Length; ++I) {
+    auto &Event = ActiveBarriersWaitList.PiEventList[I];
+    ActiveBarriers.add(Event);
+  }
+
+  pi_event Event = nullptr;
+  if (auto Res = createEventAndAssociateQueue(
+      this, &Event, PI_COMMAND_TYPE_USER, CmdList, /*IsInternal*/ true))
+    return Res;
+
+  Event->WaitList = ActiveBarriersWaitList;
+  Event->OwnZeEvent = true;
```

Comment:
It is weird that this line change is needed. Without it "internal" events are NOT being destroyed, because `OwnZeEvent` is set to false for them. Any suggestions how to improve this?

Reply:
It seems like we don't need to create this event. It is not used.

Reply:
It is used to track the "retained" dependent events to be released at the end of completed command-list.

Reply:
Ok, got it (forgot that this created event is used by command list and we track its completion).

As or the original question: it will be resolved/improved by addressing this TODO: https://github.com/intel/llvm/blob/sycl/sycl/plugins/level_zero/pi_level_zero.cpp#L3666-L3667

I will do a separate PR.

Reply:
> As or the original question: it will be resolved/improved by addressing this TODO: https://github.com/intel/llvm/blob/sycl/sycl/plugins/level_zero/pi_level_zero.cpp#L3666-L3667

this is not pointing to a TODO, can you send the correct link?

Reply:
Sorry, here: https://github.com/intel/llvm/blob/sycl/sycl/plugins/level_zero/pi_level_zero.cpp#L3677-L3678